### PR TITLE
Devel/emergency stop

### PR DESCRIPTION
**Pull request by olivefi**
Add emergency stop functionality which overrides incoming commands. The emergency stop is triggered when:
- Command exceeds RPM bound
- Command exceeds Torque bound
- Robot exceeds maximum velocity (0.3 * maximum velocity for backwards)
- Emergency stop topic contains a `std_msgs::Bool` with value `true`

### CODE REVIEW

File: roboteq_diff_driver/include/roboteq_diff_driver/driver.h [hunk 1]
```diff
@@ -32,6 +33,7 @@
 #include "rogoteq_diff_msgs/RequestOdometryCovariances.h"
 #endif
 
+#define PI 3.14159265
```
Comment (GoldenZephyr) [inline roboteq_diff_driver/include/roboteq_diff_driver/driver.h]: Please use const variables instead of macros

**Review by GoldenZephyr — CHANGES_REQUESTED** [review]
Two high-level comments:
1) Let's make these changes directly on roboteq_diff_driver/master instead of feature/IO_node, because the other branch needs a little more work and we want this merged in asap. I know that makes it a little more annoying to implement but I think it's worth it
2) I don't think it should estop the tractor if we hit the max rpm. *maybe* it can act like that in a special debug safety mode, but in general it should just clamp to some maximum RPM. I agree that we don't want the tractor to start moving arbitrarily fast, but for whatever maximum nominal RPM we choose, it's reasonable to expect the tractor is allowed to operate near that limit, and we don't want a transient above the limit to result in unexpected stopping.

**Comment by olivefi** [comment]
> Two high-level comments:
> 
> 1. Let's make these changes directly on roboteq_diff_driver/master instead of feature/IO_node, because the other branch needs a little more work and we want this merged in asap. I know that makes it a little more annoying to implement but I think it's worth it
> 2. I don't think it should estop the tractor if we hit the max rpm. _maybe_ it can act like that in a special debug safety mode, but in general it should just clamp to some maximum RPM. I agree that we don't want the tractor to start moving arbitrarily fast, but for whatever maximum nominal RPM we choose, it's reasonable to expect the tractor is allowed to operate near that limit, and we don't want a transient above the limit to result in unexpected stopping.

I've thought about it a bit, here are my considerations: 
We definitely do not want to e-stop just after touching the max velocity, but introducing feedback to reaching the max velocity is dangerous: if we send a clamped command as a reaction to reaching max velocity, we risk an infinite loop where the clamped command leads to something higher than max velocity, thus triggering the clamped command again. We would not be able to stop the robot in that case aside from killing the node.
I therefore propose splitting the parameters:
`max_rpm` and `critical_rpm`
`max_velocity` and `critical_velocity`
The `max` params can be used to check the received inputs, and clamp if needed. If we receive 20 consecutive inputs which need to be clamped, we assume the node sending us information is broken somehow and e-stop.
The `critical` params are for internal use and trigger an e-stop. This is so if something in the node breaks, we e-stop as a last resort.
Example params:
- `max_rpm` : 70
- `critical_rpm`: 100
- `max_velocity`: 1.2
- `critical_velocity`: 1.6
The rationale is that the `critical` params are chosen sufficiently above the `max` values so that they shouldn't be triggered in normal operation. Overshoot of ~10% may happen, but if we're going at 1.6 m/s something fundamental is likely wrong.

File: roboteq_diff_driver/src/driver.cpp [hunk 2]
```diff
@@ -98,39 +153,52 @@ void MainNode::rpm_setup()
...
+      ROS_INFO_STREAM("Drive mode: TORQUE");
+      controller.write("^MMOD 1 5\r");
+      controller.write("^MMOD 2 5\r");
+      controller.write("^MAC 1 50000\r"); //maximum acceleration, in torque mode this milliamps/sec
+      controller.write("^MAC 2 50000\r");
+      controller.write("^MDEC 1 50000\r"); //maximum deceleration
+      controller.write("^MDEC 2 50000\r");
+      break;
+    default: 
+      ROS_WARN_STREAM("Unknown drive mode, defaulting to KINEMATIC");
```
Comment (GoldenZephyr) [inline roboteq_diff_driver/src/driver.cpp]: I think we should error out / exit instead

**Review by GoldenZephyr — APPROVED** [review]
I made one comment about a place where I think we should crash on unknown mode instead of continuing, other than that looks good. I guess after fixing that, the next step is to merge this into dev and then move dev into `monsterduck` instead of keeping as submodule?

**Pull request merged**